### PR TITLE
osd: only call stat/getattrs once per object during deep-scrub

### PR DESCRIPTION
**Pull request by athanatos**
Fixes: https://tracker.ceph.com/issues/65185

See bug for details.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
Waiting for test branches to build.

wip-sjust-squid-testing-2024-04-18
wip-sjust-testing-2024-04-18

**Review by rzarzynski — APPROVED** [review]

**Comment by yuriw** [comment]
This PR is under test in [https://tracker.ceph.com/issues/65592](https://tracker.ceph.com/issues/65592).

**Comment by rzarzynski** [comment]
jenkins test windows

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -12426,6 +12426,7 @@ int BlueStore::getattrs(
       r = -ENOENT;
       goto out;
     }
+    aset.clear();
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/os/bluestore/BlueStore.cc:12429]: I'd be truly shocked if there were any users doing this on purpose.  I think the balance of probability is that if there actually were any users that passed in non-empty maps, we'd want this behavior anyway.

**Comment by athanatos** [comment]
https://pulpito.ceph.com/sjust-2024-04-19_02:04:26-upgrade:reef-x:stress-split-wip-sjust-squid-testing-2024-04-18-distro-default-smithi/ doesn't show the scrub error, but all of the tests failed anyway.  Is anyone looking into these other upgrade test failures?  @rzarzynski

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (athanatos) [reply src/os/bluestore/BlueStore.cc:12429]: I don't want to leave this behavior in the code base.  It's quite dangerous and I'm not really worried about backporting it tbh.

**Comment by rzarzynski** [comment]
The problem behind the first dead job in the run is failure in pulling a container image:

```
rzarzynski@teuthology:/a/sjust-2024-04-19_02:04:26-upgrade:reef-x:stress-split-wip-sjust-squid-testing-2024-04-18-distro-default-smithi/7663330$ less teuthology.log 
...
2024-04-19T02:41:02.894 DEBUG:teuthology.orchestra.run.smithi059:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:reef shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid f762b95a-fdf3-11ee-bc92-c7b262605968 -e sha1=0e14f5e6f0b65f450e578e8b6f6baaeb0179d877 -- bash -c 'ceph orch upgrade start --image quay.ceph.io/ceph-ci/ceph:$sha1'
...
2024-04-19T02:41:12.977 INFO:journalctl@ceph.mon.c.smithi059.stdout:Apr 19 02:41:12 smithi059 bash[31138]: cephadm 2024-04-19T02:41:11.008362+0000 mgr.y (mgr.24320) 186 : cephadm 1 Upgrade: First pull of quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877
...
2024-04-19T02:50:00.389 INFO:journalctl@ceph.mon.b.smithi081.stdout:Apr 19 02:50:00 smithi081 bash[23242]: cluster 2024-04-19T02:50:00.000166+0000 mon.a (mon.0) 1544 : cluster 3 [WRN] UPGRADE_FAILED_PULL: Upgrade: failed to pull target image
...
024-04-19T02:50:00.478 INFO:journalctl@ceph.mon.c.smithi059.stdout:Apr 19 02:50:00 smithi059 bash[31138]: cluster 2024-04-19T02:50:00.000193+0000 mon.a (mon.0) 1545 : cluster 3     host smithi059 `cephadm pull` failed: cephadm exited with an error code: 1, stderr: Pulling container image quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877...
2024-04-19T02:50:00.478 INFO:journalctl@ceph.mon.c.smithi059.stdout:Apr 19 02:50:00 smithi059 bash[31138]: cluster 2024-04-19T02:50:00.000218+0000 mon.a (mon.0) 1546 : cluster 3 Non-zero exit code 1 from /usr/bin/docker pull quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877
2024-04-19T02:50:00.478 INFO:journalctl@ceph.mon.c.smithi059.stdout:Apr 19 02:50:00 smithi059 bash[31138]: cluster 2024-04-19T02:50:00.000261+0000 mon.a (mon.0) 1547 : cluster 3 /usr/bin/docker: stderr Error response from daemon: manifest for quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877 not found: manifest unknown: manifest unknown
2024-04-19T02:50:00.478 INFO:journalctl@ceph.mon.c.smithi059.stdout:Apr 19 02:50:00 smithi059 bash[31138]: cluster 2024-04-19T02:50:00.000281+0000 mon.a (mon.0) 1548 : cluster 3 ERROR: Failed command: /usr/bin/docker pull quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877
...
```

**Comment by rzarzynski** [comment]
Grepped specifically for this kind of error in the two other ones:

```
rzarzynski@teuthology:/a/sjust-2024-04-19_02:04:26-upgrade:reef-x:stress-split-wip-sjust-squid-testing-2024-04-18-distro-default-smithi/7663331$ less teuthology.log
...
2024-04-19T02:50:01.030 INFO:journalctl@ceph.mon.b.smithi179.stdout:Apr 19 02:50:00 smithi179 bash[23642]: cluster 2024-04-19T02:50:00.000215+0000 mon.a (mon.0) 993 : cluster 3 ERROR: Failed command: /usr/bin/docker pull quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877
```

```
rzarzynski@teuthology:/a/sjust-2024-04-19_02:04:26-upgrade:reef-x:stress-split-wip-sjust-squid-testing-2024-04-18-distro-default-smithi/7663334$ less teuthology.log
...
2024-04-19T03:00:01.024 INFO:journalctl@ceph.mon.c.smithi031.stdout:Apr 19 03:00:00 smithi031 bash[31409]: cluster 2024-04-19T03:00:00.000187+0000 mon.a (mon.0) 1796 : cluster 3 ERROR: Failed command: /usr/bin/docker pull quay.ceph.io/ceph-ci/ceph:0e14f5e6f0b65f450e578e8b6f6baaeb0179d877
```

Everything else in red because a package problem: `yum-utils-4.5.0-1.el10.noarch from CentOS-BaseOS\ \ requires python3-dnf >= 4.19.0, but none of the providers can be installed`.

CC: @athanatos.

**Comment by ronen-fr** [comment]
I think we can merge this now. As Sam is on PTO, I'll wait a bit for possible objections (@rzarzynski ?), then probably merge

**Comment by rzarzynski** [comment]
@ronen-fr: https://tracker.ceph.com/issues/65592 tracks the QA. @yuriw, @ljflores: ping.

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65592

**Pull request merged**